### PR TITLE
Rename and split out document and page models into roles

### LINKED ISSUE #63: Use roles to reduce code in the document models
Both the PDF and CairoImageSurface have shared code. Use a role to simplify that.

### PR DESCRIPTION
**Pull request by zmughal**
Rename and split out document and page models into roles
- The renaming puts models into a hierarchy that
  groups the models together by type (either a document or a page).
- Splits out common functionality shared among the documents and pages
  to roles.
- Move the PDF document test from the `xt/` directory to the `t/` directory
  so that it is part of the main test suite.

Closes #63.

### CODE REVIEW

**Comment by zmughal** [comment]
@project-renard/dev, this is ready for review.

**Comment by zmughal** [comment]
The identifiers in #70 will need to change to accommodate this change and quite possibly more PRs.

File: lib/Renard/Curie/Model/Document/Role/HasFilename.pm [hunk 1]
```diff
@@ -0,0 +1,13 @@
+use Modern::Perl;
+package Renard::Curie::Model::Document::Role::HasFilename;
```
Comment (zmughal) [inline lib/Renard/Curie/Model/Document/Role/HasFilename.pm]: Rename to `::FromFile`

**Comment by StayerX** [comment]
One of the unit tests is failing. Very god idea. I like it. I think that as you already said there will be some overlap in code this way. Does Perl support interfaces. I think there should be a general interface that the rest should inherit in this case.

**Comment by zmughal** [comment]
That failing check is because the coverage decreased. I'm not sure if that has to do with fewer lines being tested or the fact that this refactoring has reduced the duplicated code.

Yes, there are interfaces. Under the hood, Perl has multiple inheritance, but to make multiple inheritance work in a consistent way, the [`Moo`](http://p3rl.org/Moo) and [`Moose`](http://p3rl.org/Moose) object systems use something called Roles (these are known as Traits in Smalltalk and Scala) to allow for including implementation inside of the interfaces.

Take a look at this documentation:
- [Intro to Moose](https://rawgit.com/moose/intro-to-moose/master/slides/index.html) slides
- Slides for development of [App::MathTutor](http://blogs.perl.org/users/jens_rehsack/2014/01/moo-in-practice---appmathtutor.html)
- [Moose::Manual](https://metacpan.org/pod/distribution/Moose/lib/Moose/Manual.pod)

I agree that the `Document` and `Page` should have a general interface. I'm not sure what should go in there yet until we have more types of implementation classes.

**Comment by zmughal** [comment]
More about Roles/Traits:

Roles are an example of https://en.wikipedia.org/wiki/Composition_over_inheritance

They don't have the drawback of Interfaces (like the ones in Java) because you can have implemented methods.

And at the same time, they don't have the drawbacks of Mixins (created through Multiple Inheritance) because they can check for conflicting names.
- https://en.wikipedia.org/wiki/Role-oriented_programming
- https://en.wikipedia.org/wiki/Trait_(computer_programming)
- http://scg.unibe.ch/research/traits/

**Comment by ChiragGhanshani** [comment]
@zmughal, It looks like this is just restructuring the project. I'm not sure why that would reduce the coverage. Is there a coverage report generated by running tests? We could use it to see which lines of code were missed.

**Comment by zmughal** [comment]
Coveralls shows the coverage and line numbers. I will show you all how to generate the reports locally with Devel::Cover when I get home tonight. Unless you want to read up on it yourself.

**Comment by ChiragGhanshani** [comment]
it'll be easier if you showed us because I have to work on SDP and DB today

**Comment by zmughal** [comment]
Sure thing. I will add a document with screenshots to the wiki.

**Comment by zmughal** [comment]
Will need to resolve the merge conflicts. Going to rebase and diff now.

**Comment by zmughal** [comment]
It appears that there is an issue with the way the tests are running on Travis-CI. It looks like the PDF tests are not running properly? I will go more in-depth later.

In the meantime, I wrote a script to run the coverage locally and it is currently in a PR at https://github.com/project-renard/devops/pull/8. To run, do the following:

``` shell
. ../../devops/devops/ENV.sh
renard_run_cover_on_branch 63-document-model-roles
renard_run_cover_on_branch master
```

It appears that the coverage has actually improved:

master:
curie-63-coverage-master

63-document-model-roles:
curie-63-coverage-63

**Comment by zmughal** [comment]
I removed the t/ directories from the coverage to get something more realistic and got:
- 63-document-model-roles: 66.3%

xming_2016-04-09_16-17-31
- master: 63.4%

xming_2016-04-09_16-17-46

**Comment by zmughal** [comment]
Haha, the reason why the coverage dropped was because I did the following rename:

```
0 xt/Renard/Curie/Model/PDFDocument.t → xt/Renard/Curie/Model/Document/PDF.pm
```

When it has a `.pm` extension, it does not get used as a test. I'll fix this. In fact, now it should go under the `t/` directory instead of the `xt/` directory because we can skip it due to the changes that added a testing library in #70.

**Comment by zmughal** [comment]
@project-renard/dev Look good to merge now?

I've only fixed a couple issues in the documentation and renamed the file as I said in the comment https://github.com/project-renard/curie/pull/71#issuecomment-207947077.

**Comment by zmughal** [comment]
@project-renard/dev Ping!

**Comment by StayerX** [comment]
LGTM :+1:

**Comment by zmughal** [comment]
@ChiragGhanshani .

**Comment by zmughal** [comment]
Since @ChiragGhanshani is busy right now, I think we should merge to continue the momentum.

**Pull request merged**